lib/UnexpectedMitmMocker.js

```diff
@@ -0,0 +1,538 @@
+
+            match = chunkString.match(/^Transfer-Encoding: /m);
+            if (match) {
+                sawHeaders = true;
+            }
+
+            match = chunkString.match(/^Content-Length: (\d+)/m);
+            if (match) {
+                // handle the current chunk containing body bytes
+                bodyByteCount = calculateBodyByteCount(chunk);
+                contentLengthByteCount = +match[1];
+                hasContentLength = true;
+                sawHeaders = true;
+            }
+        } else {
+            bodyByteCount += chunk.length;
+        }
+
+        if ((isTrailerChunk(chunk) && !hasContentLength) ||
+                (hasContentLength && bodyByteCount === contentLengthByteCount)) {
+            /*
+             * explicitly execute the callback returning raw data here
+             * to ensure it is run before issuing the final write which
+             * will complete the request on the wire
+             */
+            callback(null, Buffer.concat(rawChunks));
+        }
+
+        _write.call(connection, chunk, encoding, cb);
+    };
+}
+
+function attachResponseHooks(res, callback) {
+    var hasEnded = false; // keep track of whether res.end() has been called
+
+    attachConnectionHook(res.connection, function (err, rawBuffer) {
+        if (hasEnded) {
+            callback(null, rawBuffer);
+        }
+    });
+
+    ['end', 'destroy'].forEach(function (methodName) {
+        var orig = res[methodName];
+        res[methodName] = function (chunk, encoding) {
+            if (methodName === 'end') {
+                hasEnded = true;
+            }
+
+            var returnValue = orig.apply(this, arguments);
+
+            if (methodName === 'destroy') {
+                hasEnded = true;
+                callback(null);
+            }
+
+            // unconditionally return value given we do not wrap write
+            return returnValue;
+        };
+    });
+}
+
+function observeResponse(res) {
+    return expect.promise(function (resolve, reject) {
+        attachResponseHooks(res, function (err, rawBuffer) {
+            resolve(rawBuffer);
+        });
+    });
+}
+
+function trimMockResponse(mockResponse) {
+    var responseProperties = {};
+    responseProperties.statusCode = mockResponse.statusCode;
+    responseProperties.statusMessage = mockResponse.statusMessage;
+    responseProperties.protocolName = mockResponse.protocolName;
+    responseProperties.protocolVersion = mockResponse.protocolVersion;
+    responseProperties.headers = formatHeaderObj(mockResponse.headers.valuesByName);
+
+    if (mockResponse.body) {
+        // read out the messy decoded body
+        responseProperties.body = mockResponse.body;
+    }
+
+    return responseProperties;
+}
+
+
+function UnexpectedMitmMocker(options) {
+    this.descriptions = options.descriptions || [];
+    this.timeline = null;
+    this.fulfilmentValue = null;
+
+    var that = this;
+    Object.defineProperty(this, 'isComplete', {
```

Comment:
Couldn't this live in `UnexpectedMitmMocker.prototype`?

Reply:
IIRC it's actually not used right now - my temptation would be to remove it until necessary (for some additional context, this was an early version of what lead to the completion marker).

Reply:
Even better :)